### PR TITLE
Perf protocol

### PR DESCRIPTION
**Pull request by MarcoPolo**
Implements https://github.com/libp2p/specs/pull/478

### CODE REVIEW

File: p2p/protocol/perf/perf.go [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+	_, err := io.ReadFull(s, u64Buf)
+	if err != nil {
+		log.Errorw("err", err)
+		s.Reset()
+		return
+	}
+
+	bytesToSend := binary.BigEndian.Uint64(u64Buf)
+
+	_, err = p.drainStream(context.Background(), s, buf)
```
Comment (marten-seemann) [inline p2p/protocol/perf/perf.go:48]: This doesn't seem very useful, see my comment on the spec: https://github.com/libp2p/specs/pull/478/files#r1122634411

File: p2p/protocol/perf/perf.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	err = p.sendBytes(context.Background(), s, bytesToSend, buf)
+	if err != nil {
+		log.Errorw("err", err)
+		s.Reset()
+		return
+	}
+
+}
+
+func (ps *PerfService) sendBytes(ctx context.Context, s network.Stream, bytesToSend uint64, buf []byte) error {
```
Comment (marten-seemann) [inline p2p/protocol/perf/perf.go:64]: Easier to read if you allocate `buf` in this function.

File: p2p/protocol/perf/perf.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+	}
+	s.CloseWrite()
+
+	return nil
+}
+
+func (ps *PerfService) drainStream(ctx context.Context, s network.Stream, buf []byte) (uint64, error) {
+	var recvd uint64
+	for {
+		n, err := s.Read(buf)
```
Comment (marten-seemann) [inline p2p/protocol/perf/perf.go:85]: You can copy into io.Discard.

File: p2p/protocol/perf/perf.go (hunk 1 above)
Comment (marten-seemann) [reply p2p/protocol/perf/perf.go:48]: Misunderstanding on my part. This is fine.

**Comment by mxinden** [comment]
With https://github.com/libp2p/test-plans/pull/184 merged, we now have a libp2p-perf protocol implementation on top of go-libp2p in `libp2p/test-plans`. Thus this pull request can be closed @MarcoPolo, right?

**Pull request closed without merging**